core/src/main/scala/org/apache/spark/Dependency.scala

```diff
@@ -88,14 +90,53 @@ class ShuffleDependency[K: ClassTag, V: ClassTag, C: ClassTag](
   private[spark] val combinerClassName: Option[String] =
     Option(reflect.classTag[C]).map(_.runtimeClass.getName)
 
-  val shuffleId: Int = _rdd.context.newShuffleId()
+  val shuffleId: Int = if (isContinuous) {
+    // This will not be reset in continuous processing, set an invalid value for now.
+    Int.MinValue
+  } else {
+    _rdd.context.newShuffleId()
+  }
 
-  val shuffleHandle: ShuffleHandle = _rdd.context.env.shuffleManager.registerShuffle(
-    shuffleId, _rdd.partitions.length, this)
+  val shuffleHandle: ShuffleHandle = if (isContinuous) {
+    null
+  } else {
+    _rdd.context.env.shuffleManager.registerShuffle(
+      shuffleId, _rdd.partitions.length, this)
+  }
 
-  _rdd.sparkContext.cleaner.foreach(_.registerShuffleForCleanup(this))
+  if (!isContinuous) {
+    _rdd.sparkContext.cleaner.foreach(_.registerShuffleForCleanup(this))
+  }
 }
 
+/**
+ * :: DeveloperApi ::
+ * Represents a dependency on the output of a shuffle stage of continuous type.
+ * Different with ShuffleDependency, the continuous dependency only create on Executor side,
+ * so the rdd in param is deserialized from taskBinary.
+ */
+@DeveloperApi
+class ContinuousShuffleDependency[K: ClassTag, V: ClassTag, C: ClassTag](
+    rdd: RDD[_ <: Product2[K, V]],
+    dep: ShuffleDependency[K, V, C],
+    continuousEpoch: Int,
+    totalShuffleNum: Int,
+    shuffleNumMaps: Int)
+  extends ShuffleDependency[K, V, C](
+    rdd,
+    dep.partitioner,
+    dep.serializer,
+    dep.keyOrdering,
+    dep.aggregator,
+    dep.mapSideCombine, true) {
+
+  val baseShuffleId: Int = dep.shuffleId
+
+  override val shuffleId: Int = continuousEpoch * totalShuffleNum + baseShuffleId
```

Comment:
We surely can't artificially construct shuffle IDs like this. IIUC only shuffle IDs returned from SparkContext.newShuffleId() are valid shuffle IDs.

Reply:
Sorry, I'm not sure I understand. The SparkContext is responsible for assigning shuffle IDs, so it's not valid to register a ContinuousShuffleDependency with a shuffle ID which SparkContext did not assign.

Reply:
Got it, if we move EpochCoordinator to SparkEnv, I think we can re-implement the shuffle register on driver side, totally controlled by EpochCoordinator. Even if the EpochCoordinator doesn't manage the shuffle register work, I think EpochCoordinator should move to SparkEnv and take more work in shuffle support in your design, am I right?